source/reduce/remove_resource_variable_reduction_opportunity.cpp

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/reduce/remove_resource_variable_reduction_opportunity.h"
+
+#include "source/opt/ir_context.h"
+
+namespace spvtools {
+namespace reduce {
+
+bool RemoveResourceVariableReductionOpportunity::PreconditionHolds() {
+  return true;
+}
+
+void RemoveResourceVariableReductionOpportunity::Apply() {
+  assert(variable_->opcode() == SpvOpVariable &&
+         "The instruction being removed should be OpVariable.");
+  assert(variable_->GetSingleWordInOperand(0) != SpvStorageClassFunction &&
+         "The variable must not have function storage class.");
+
+  std::vector<uint32_t> use_ids_to_kill;
+
+  // Find all descriptor set / binding decorations that target the variable.
+  context_->get_def_use_mgr()->ForEachUse(
+      variable_,
+      [&use_ids_to_kill](const opt::Instruction* user, uint32_t /*use_index*/) {
+        assert(user->opcode() == SpvOpDecorate &&
+               "The only remaining uses should be decorations.");
+        assert(IsAcceptableDecoration(*user) &&
+               "Only very specific decorations should target the variable.");
+        use_ids_to_kill.push_back(user->result_id());
+      });
+
+  // Kill each of these decorations.
+  for (auto id : use_ids_to_kill) {
+    context_->KillInst(context_->get_def_use_mgr()->GetDef(id));
+  }
+
+  // Kill the variable.
+  context_->KillInst(variable_);
```

Comment:
I think this already kills the corresponding names and decorations. Also, it may do so more carefully than the above.

Reply:
Good catch - and by exploiting that fact, this reduction opportunity becomes identical to remove_instruction_reduction_opportunity - i.e., there's nothing special to be done.  I have simplified the PR on that basis.